IO: separate pack and set commands

DougBurke opened the pull request:
# Summary

Replace the logic of the packup argument to the set I/O commands by adding new pack_xxx_data commands (which return an object, and so correspond to the old packup=True case) and make the existing set_xxx_data calls never return anything (the old packup=False behaviour). As part of this, ensure that the dummy backend provides arguments and - for most arguments - types, which means the dummy backend can be thought of as defining the backend interface. There should be little change to users who use the sherpa.astro.io module routines.


# Details

Required
- #1926 

Followed by
- #1921 

A clean up of the crates and pyfits backend modules, as well as the general backend. These changes generally should not change behaviour, but there may be some corner cases where it may make a difference (e.g. what is the name of a FITS HDU for a "generic" table). Although there are a number of additional changes that can be made to the backend code, this is planned for a later PR (#1921), with the changes here being enough to make it clearer (to me) what the code is doing and what changes we can make in the future (the idea being that we can move much of the logic into the generic layer, and remove the special-case behavior, particularly for the OGIP file types, from the backend code). The changes here include

- pylint suggestions
- add typing for some of the routines to make it clearer what is going on
- move to f-strings except for logging statements, where we want to use "lazy formatting"
- remove some code that appears to have been either "wishful thinking" or is no-longer relevant (normally this is about removing some special-case behaviour we can't trigger in our tests)
- crates:
  - simplify the recent code added in 4.16.0 to create the TLMIN value fo the F_CHAN column of RMF files with the crates backend (there is a simpler way to do it than the low-level use of cxcdm added in PR #1836)
  - simplify or remove some of the wrapper code as we can just create the dataset objects directly rather than going through a wrapper
  - generally try to have the "get a column/key" routine logic be in the "require" version and then the "try" version catches the error and returns a None (rather than have the "require" version call the "try" version and then error out if None is returned)
- pyfits:
  - the internal open routine, which takes a string of a fits.HDUList, did not record whether the file was opened, and hence should be closed, which has now been fixed (to better avoid getting "buffer still open" errors on exit of the interpreter)
  - simplify the wrapper code as a number of cases appear to no-longer be relevant or could never happen

The dummy backend can now be thought of as defining the API for a general backend (e.g. there is now an indication of both the input and output from all the routines). We could make this into a class, which could help, but to be honest is it worth it? We aren't likely to add a new backend any time soon, or want to sub-class either of the existing backends, so I am not convinced moving to a class will improve the code in any way.

The final commit changes the backend code to add pack_xxx_data routines and to drop the packup argument from the set_xxx_data calls. The idea is that the set_xxx_data call can just call pack_xxx_data and then write out the response but it's not always done like this (since we can go via the set_table_data call, for example). However, the individual pack routines are more used in the follow-up PR #1921.

DougBurke commented:
Rebased to pick up #1938

hamogu reviewed (changes requested):
I'm sorry this fell of my radar. It was too long to do it right away, and then I forgot. Most changes I suggest are cosmetic, but if we touch it anyway, I want to strongly advocate to make arrays using numpy where possible.
```
n = 1234
%timeit arr = np.array([None] * n)
74.9 µs ± 2.45 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
%timeit arr = np.full(n, None)
11.1 µs ± 92.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

hamogu commented:
In general, I agree that this module-based way of writing the backends is not ideal, but not worth changing at this point. One could use classes with inheritance or protocols or entry points and that might be intellectually satisfying, but it's not worth the effort here.

DougBurke commented on sherpa/astro/io/crates_backend.py:
```diff
@@ -166,51 +182,55 @@ def _try_key(crate, name, dtype=str):
-            if type(val) is numpy.str_ and val == '':
-                val = ''
+        # empty numpy strings are not recognized by load pickle!
+        if isinstance(val, np.str_) and val == '':
+            val = ''
```
This is old code and I've been wary of changing it too much. It looks like it's trying to convert from a NumPy string to a "normal" one, but why I don't know (I know there have been historical issues with passing data around, for a variety of reasons). So this is one area I decided not to touch (I am running tests with a crates backend but I'm slightly careful as we don't have a CI run with it). 

I can look again at this code, as it's been a while since I last worked on it.

DougBurke replied:
This is very-much a "I'm sure there was a reason for this, but I don't know what". As we don't have a test that exercises this I am a little reluctant to change it at this time (which is somewhat funny giving the sweeping changes I'm making to the rest of the I/O code here...).

DougBurke commented on sherpa/astro/io/crates_backend.py:
```diff
@@ -307,99 +318,121 @@ def _try_key_list(crate, keyname, num, dtype=SherpaFloat, fix_type=False):
     # Make a copy of the data, since we don't know that pycrates will
     # do something sensible wrt reference counting
-    key = numpy.array([key.value] * num)
+    key = np.array([key.value] * num)
```
There's always the balance between speed and convenience (e.g. I don't **myself** think that the I/O routines are time critical, but I tend not to be using large files). 

Saying that, I'm not sure that `[foo] * num` is more or less reasable than using `np.full`!

hamogu replied:
I actually benchmarked that. For this one line, the different in runtime is about a factor of 5-10 in speed for a column with 1024 entries and bigger for bigger files. It's still >> 1s so it's certainly not critical because I/O routines tend to be limited by disk I/O anyway. I would not open a new PR just for this. However, "sherpa performance" was on my SDS task list for 2023 and is again for 2024, so I figured I point it out when the code is touched anyway and it's just an easy change to do that is about equally readable.

hamogu replied:
I still want the `np.fill...` here...

hamogu replied:
and below where we have the same pattern.

DougBurke commented on sherpa/astro/io/crates_backend.py:
```diff
@@ -1047,26 +1066,28 @@ def get_pha_data(arg, make_copy=True, use_background=False):
+            counts = _require_col_list(pha, 'COUNTS',
+                                       make_copy=make_copy, fix_type=True)
+
         else:
             if not pha.column_exists('RATE'):
                 raise IOErr('reqcol', 'COUNTS or RATE', filename)
```
Gah. This is one of the areas I was "concentrating" on (thanks, in part, to `pylint`) but obviously missed this one.

DougBurke commented on sherpa/astro/io/pyfits_backend.py:
```diff
@@ -434,7 +404,8 @@ def get_header_data(arg, blockname=None, hdrkeys=None):
 
     finally:
-        tbl.close()
+        if close:
+            tbl.close()
```
I'll have a look. I am not particularly a fan of having to pass the "close" flag around, so if we can avoid it all the better.

DougBurke commented on sherpa/astro/io/pyfits_backend.py:
```diff
@@ -1081,19 +1078,19 @@ def req_sfloat(key):
             counts = try_sfloat("COUNTS")
             staterror = _try_vec(hdu, 'STAT_ERR', size=num)
-            if numpy.equal(counts, None).any():  # _try_vec can return an array of Nones
+            if np.equal(staterror, None).any():
```
There's some funky stuff going on with this (I don't think we have good test coverage of this) so I'll have a look.

DougBurke replied:
Not sure what this was meant to be doing - I think it was a typo by me - but it lead to a long sequence of changes to try and work out why just the python 3.11 build was failing on CI. I haven't worked out why, but #2017 (which is a re-worked version of this) now has this fixed.

hamogu commented on sherpa/astro/io/crates_backend.py:
```diff
@@ -242,38 +262,25 @@ def _require_key(crate, name, dtype=str):
+             make_copy: bool = False,
+             fix_type: bool = False,
+             dtype: type = SherpaFloat) -> Optional[np.ndarray]:
+    """Return the column data or return None."""
```
I was just suggesting to add "is the column does not exist" to the docstring, but GH seemed to have changed their website and I could not make "suggestions" when I reviewed this so I wrote it in plain text. Sorry that was not clear. (Now the button is there again.)

I agree that it would be *nice* to unify the backend code and remove or combine some of those utility functions, but is that the best use of our time?

DougBurke commented on sherpa/astro/io/pyfits_backend.py:
```diff
@@ -1490,32 +1489,45 @@ def set_image_data(filename, data, header, ascii=False, clobber=False,
+    try:
+        size = len(args[0])
+    except (TypeError, IndexError) as exc:
+        raise IOErr('noarrayswrite') from exc
```
There's also some column-selection code (e.g. what columns to look for when not given explicitly) that would likely benefit from this sort of approach.

DougBurke commented:
@hamogu - I've tried switching to a class-based backend approach - based on the plotting approach but a lot simpler as I don't see the benefit (instead, this relies on the backend module containing a `Backend` class which is then used). There's some fall out, but mainly in the test code which has to decide what to do when given a particular backend.

There is also some small changes in behaviour, but these are generally "small" and not-often used cases, so hopefully it's not a problem (e.g. we now create a header line when writing out arrays with the pyfits backend when we didn't before).

I did address some comments, but not all of them because some of them are better done in the next PR (#1921) which is going to need quite a rework (even if we decide the backend-class changes aren't worth it).

hamogu commented:
@DougBurke What is your opinion for the class-based design then? I'm soso, and certainly did not want to leave the impression that "I require classes to approve this PR". I could see some advantages, but it's really minor. You obviously felt it's worth your time or you would not have implemented it, but your comments sound less then enthusiastic if read at face value. Or should I take that as British understatement?

hamogu commented on sherpa/astro/io/__init__.py:
```diff
@@ -1046,24 +1188,34 @@ def convert(column, dtype):
-        if numpy.issubdtype(vals.dtype, numpy.integer):
-            vals = vals.astype(numpy.int32)
+        if vals is None:
+            # Need to appease mypy
```
You choose your masters.

hamogu reviewed (commented):
I see that pyfits is converted to a class, but crates is not. Did you manybe forget to commit and push the crates backend?

Otherwise, I think this is good. Again, I don't want to push you to implement something that you don't see as a good use of your time. I think the classes will help to abstract out some common logic which will help us in the long run but it's not clear how long it will be until that investment pays off -maybe never. Since you did the work already, I fully support going forward, but only as far as you think it's useful.

DougBurke commented on sherpa/astro/io/pyfits_backend.py:
```diff
@@ -1624,19 +680,1007 @@ def _create_table_hdu(hdu: TableHDU) -> fits.BinTableHDU:
+
+        def arraycol(name):
+            vals = matrix_data[name]
+            if name == "MATRIX": print(vals)
```
yes. definitely something that shouldn't be there

DougBurke commented on sherpa/astro/io/tests/test_io_img.py:
```diff
@@ -25,12 +25,13 @@
 def backend_is(name):
     """Are we using the given backend?"""
-    return io.backend.__name__ == f"sherpa.astro.io.{name}_backend"
+    return io.backend.name == name
```
It's a close decision. It is used a lot, but it's also a really-small piece of code so is it worth it? We probably have something similar with the plot backend that perhaps could (or already is?) wrapped up in `conftest.py`.

DougBurke commented:
For https://github.com/sherpa/sherpa/pull/1929#issuecomment-1931015494 - I haven't yet decided whether I think the class-based design is better or worse or just different!

DougBurke commented:
With regards to crates, I see the `class Backend` in `crates_backend.py` in https://github.com/sherpa/sherpa/pull/1929/commits/a8a557f8dd00e94437eabaabcff95ad3a86ee5dd so I'm not sure what's going on.

hamogu commented on sherpa/astro/io/crates_backend.py:
```diff
@@ -20,59 +20,78 @@
     warning('failed to import WCS module; WCS routines will not be '
             'available')
 
 __all__ = ('get_table_data', 'get_header_data', 'get_image_data',
            'get_column_data', 'get_ascii_data',
            'get_arf_data', 'get_rmf_data', 'get_pha_data',
+           'pack_table_data', 'pack_image_data', 'pack_pha_data',
+           'pack_arf_data', 'pack_rmf_data', 'pack_hdus',
            'set_table_data', 'set_image_data', 'set_pha_data',
            'set_arf_data', 'set_rmf_data', 'set_hdus')
```
```suggestion
__all__ = ('Backend', )
```

hamogu commented:
> With regards to crates, I see the class Backend in crates_backend.py in https://github.com/sherpa/sherpa/commit/a8a557f8dd00e94437eabaabcff95ad3a86ee5dd so I'm not sure what's going on.

I just got confused by the `__all__` - I added a suggestion.

DougBurke commented:
Rebased because of #1952 and addressed some review comments by @hamogu. I still haven't decided whether the use-a-class-for-the-backend code is worth it or not. If not I probably want to copy over some of the documentation improvements.

DougBurke commented on sherpa/astro/io/pyfits_backend.py:
```diff
@@ -540,47 +512,57 @@ def get_image_data(arg, make_copy=False):
+
+            # proper calculation of crpix wrt PHYSICAL coords
+            if (crpixw is not None and crvalp is not None and
+                cdeltp is not None and crpixp is not None):
```
Unfortunately the values are `Optional[np.ndarray]` which means that we then get errors because of this - e.g.

```
103 >>> from sherpa.astro.io import read_image
104 >>> img = read_image(data_dir + 'image2.fits')
UNEXPECTED EXCEPTION: ValueError('The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()')
Traceback (most recent call last):
  File "/lagado2.real/local/anaconda/envs/sherpa-xspec/lib/python3.11/doctest.py", line 1355, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest sherpa.astro.data[1]>", line 1, in <module>
  File "/lagado.real/sherpa/sherpa-xspec/sherpa/astro/io/__init__.py", line 389, in read_image
    data, filename = backend.get_image_data(arg)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lagado.real/sherpa/sherpa-xspec/sherpa/astro/io/pyfits_backend.py", line 580, in get_image_data
    if None not in [cdeltw, cdeltp]:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

DougBurke replied:
[this is for #2017 but I wanted to comment here where the issue was raised]

DougBurke commented:
Development of this is now in #2017 so I'm closing this one.
